curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -0,0 +1,855 @@
+            return;
+        }
+    }
+}
+
+int TopologyManager::CreatePartition(FsIdType fsId, const uint32_t count,
+                                     ptrf *partitionInfoList) {
+    while (partitionInfoList->size() < count) {
+        CopySetInfo copyset;
+        if (!topology_->GetAvailableCopyset(&copyset)) {
+            LOG(WARNING) << "Have no available copyset when create partition.";
+            if (kTopoErrCodeSuccess == CreateCopyset()) {
+                CreatePartition(fsId, count, partitionInfoList);
+                break;
+            } else {
+                LOG(ERROR) << "Create copyset failed when create partition.";
+                return kTopoErrCodeGenCopysetErr;
+            }
+        }
+
+        PartitionIdType partitionId = topology_->AllocatePartitionId();
+        uint64_t idStart = copyset.GetPartitionNum() *
+                        option_.IdNumberInPartition + 1;
+        uint64_t idEnd = (copyset.GetPartitionNum() + 1) *
+                        option_.IdNumberInPartition;
+        PoolIdType poolId = copyset.GetPoolId();
+        CopySetIdType copysetId = copyset.GetId();
+        std::set<MetaServerIdType> copysetMembers = copyset.GetCopySetMembers();
+
+        std::set<std::string> copysetMemberAddr;
+        for (auto item : copysetMembers) {
+            MetaServer metaserver;
+            if (topology_->GetMetaServer(item, &metaserver)) {
+                std::string addr = metaserver.GetInternalHostIp() + ":" +
+                                std::to_string(metaserver.GetInternalPort());
+                copysetMemberAddr.emplace(addr);
+            } else {
+                LOG(WARNING) << "Get metaserver info failed.";
+            }
+        }
+
+        // get leader
+        std::string leader;
+        FSStatusCode retcode = metaserverClient_->GetLeader(poolId, copysetId,
+                                                    copysetMemberAddr, &leader);
+        if (FSStatusCode::NOT_FOUND == retcode) {
+            LOG(WARNING) << "Get leader failed when create partition.";
+            continue;
+        }
+
+        retcode = metaserverClient_->CreatePartition(fsId, poolId, copysetId,
+                                        partitionId, idStart, idEnd, leader);
+        Partition partition(fsId, poolId, copysetId, partitionId,
+                            idStart, idEnd);
+        if (FSStatusCode::OK == retcode) {
+            int ret = topology_->AddPartition(partition);
+            if (kTopoErrCodeSuccess == ret) {
+                PartitionInfo *info = partitionInfoList->Add();
+                info->set_fsid(fsId);
+                info->set_poolid(poolId);
+                info->set_copysetid(copysetId);
+                info->set_partitionid(partitionId);
+                info->set_start(idStart);
+                info->set_end(idEnd);
+            } else {
+                LOG(ERROR) << "Add partition failed after create partition."
```

Comment:
delete partitions from metaserver

Reply:
> delete partitions from metaserver

This will involve some error handling, add a TODO first.